### PR TITLE
Adds support for searching user type in user creation form

### LINKED ISSUE #7326: Issue in finding the value in options tab
**Describe the bug**
There was difficulty in finding the suitable user type from the options provided because there was large number of options were available.

Screenshot 2024-03-01 at 12 58 11 AM

### PR DESCRIPTION
**Pull request by Ashutosh0602**
## Proposed Changes
There was difficulty in finding the suitable user type from the options provided because there was large number of options were available.

So, added a search tab in User Type to find out the best desired type easily from the detailed list. 
Fixed some type error and UI flaws.

- close #7326  

@coronasafe/care-fe-code-reviewers @coronasafe/code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.coronasafe.network/coronasafe-care-documentation/architecture/architecture-and-layering-of-care).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

### CODE REVIEW

File: src/Components/Users/UserAdd.tsx [hunk 1]
```diff
@@ -266,6 +267,17 @@ export const UserAdd = (props: UserProps) => {
...
 
+  const { loading: isStateLoading } = useQuery(routes.statesList, {
+    onResponse: (result) => {
+      if (!result || !result.res || !result.data) return;
+      if (userIndex <= USER_TYPES.indexOf("StateAdmin")) {
+        setStates([authUser.state_object!]);
+      } else {
+        setStates(result.data.results);
+      }
+    },
```
Comment (rithviknishad) [inline src/Components/Users/UserAdd.tsx]:
```suggestion
    onResponse: ({ data }) => {
      if (!data) {
        return;
      }
      
      if (userIndex <= USER_TYPES.indexOf("StateAdmin")) {
        setStates([authUser.state_object!]);
      } else {
        setStates(data.results);
      }
    },
```

File: src/Components/Users/UserAdd.tsx [hunk 2]
```diff
@@ -597,14 +594,18 @@ export const UserAdd = (props: UserProps) => {
...
+    return <Loading />;
+  }
+
   return (
     <Page
       title={headerText}
       options={
         <Link
-          href="https://school.coronasafe.network/targets/12953"
+          href="https://school.coronasafe.network/targets/12953 "
```
Comment (rithviknishad) [inline src/Components/Users/UserAdd.tsx:609]: Why was this changed?

```suggestion
          href="https://school.coronasafe.network/targets/12953"
```

File: src/Components/Users/UserAdd.tsx [hunk 3]
```diff
@@ -597,14 +594,18 @@ export const UserAdd = (props: UserProps) => {
...
           className="inline-block rounded border border-gray-600 bg-gray-50 px-4 py-2 text-gray-600 transition hover:bg-gray-100"
-          target="_blank"
+          target="_blank "
```
Comment (rithviknishad) [inline src/Components/Users/UserAdd.tsx:611]: Why was this changed?
```suggestion
          target="_blank"
```

File: src/Components/Users/UserAdd.tsx [hunk 4]
```diff
@@ -631,15 +632,26 @@ export const UserAdd = (props: UserProps) => {
                 showAll={false}
               />
             </div>
-            <SelectFormField
+
+            {/* <SelectFormField
```
Comment (rithviknishad) [inline src/Components/Users/UserAdd.tsx]: You can remove this instead of commenting it

File: src/Components/Users/UserAdd.tsx [hunk 5]
```diff
@@ -631,15 +632,26 @@ export const UserAdd = (props: UserProps) => {
...
               optionLabel={(o) => o.role + (o.readOnly ? " (Read Only)" : "")}
               optionValue={(o) => o.id}
-            />
+            /> */}
 
+            <AutocompleteFormField
+              {...field("user_type")}
+              label="User Type"
+              placeholder="Search for a user type..."
+              // value={userTypes}
```
Comment (rithviknishad) [inline src/Components/Users/UserAdd.tsx]:
```suggestion
```

File: src/Components/Users/UserAdd.tsx [hunk 6]
```diff
@@ -326,7 +327,7 @@ export const UserAdd = (props: UserProps) => {
 
   const setFacility = (selected: FacilityModel | FacilityModel[] | null) => {
     setSelectedFacility(selected as FacilityModel[]);
-    const form = { ...state.form };
+    const form: any = { ...state.form };
```
Comment (rithviknishad) [inline src/Components/Users/UserAdd.tsx]: Casting to `any` not allowed.

```suggestion
    const form = { ...state.form };
```

**Review by rithviknishad — CHANGES_REQUESTED** [review]

**Comment by rithviknishad** [comment]
Also, it'd be great if you could give a meaningful title for the PR. It could convey what this PR has done. For example: "Adds support for searching user type in user creation form"

**Comment by rithviknishad** [comment]
Also, ensure cypress tests are passing

File: src/Components/Users/UserAdd.tsx (hunk 2 above)
Comment (Ashutosh0602) [reply src/Components/Users/UserAdd.tsx:609]: While saving the file extra space removed.
Is this making any conflict?

File: src/Components/Users/UserAdd.tsx (hunk 4 above)
Comment (Ashutosh0602) [reply src/Components/Users/UserAdd.tsx]: I was checking the everything was working properly I will remove and make new commit.

**Comment by Ashutosh0602** [comment]
@rithviknishad I fix the previous mentioned changes.

**Comment by rithviknishad** [comment]
Are you sure the cypress tests are passing?

**Comment by Ashutosh0602** [comment]
No, 3 out of 7 cypress test failed and rest is passed.
And currently I am trying to figure it out. If you can give some suggestion related to cypress I am new to using cypress

**Comment by Ashutosh0602** [comment]
@rithviknishad Is there something I have to do here?

**Comment by rithviknishad** [comment]
Hey, you can connect your FE to your local backend and run the cypress tests by following the documentation here...

https://github.com/coronasafe/care_fe?tab=readme-ov-file#-run-cypress-tests

Once you've identified which tests are failing, you can update the corresponding cypress tests. The test code would be present in the same file name as the cypress test spec name.
